File: web/src/services/api/saved.ts
import { api } from './client'

export type SavedItem = {
  id: string
  target_type: string
  target_id: string
  created_at: string | null
}

export type SavedListResponse = {
  items: SavedItem[]
  total: number
}

export type SavedId = {
  target_type: string
  target_id: string
}

export const savedApi = {
  list: () => api.get<SavedListResponse>('/api/v1/saved'),
  listIds: () => api.get<SavedId[]>('/api/v1/saved/ids'),
  add: (target_type: 'FAQ' | 'DISCUSSION' | 'QUESTION' | 'REPLY', target_id: string) =>
    api.post<SavedItem>('/api/v1/saved', { target_type, target_id }),
  remove: (target_type: string, target_id: string) =>
    api.delete<void>(`/api/v1/saved/${target_type}/${target_id}`),
}

File: backend/app/schemas/saved.py
from datetime import datetime
from pydantic import BaseModel, Field, ConfigDict

from app.schemas.base import BaseSchema


class SavedCreate(BaseModel):
    target_type: str = Field(..., pattern="^(FAQ|DISCUSSION|QUESTION|REPLY)$")
    target_id: str


class SavedResponse(BaseSchema):
    id: str
    target_type: str
    target_id: str
    created_at: datetime | None = None

    model_config = ConfigDict(from_attributes=True)


class SavedListResponse(BaseModel):
    items: list[SavedResponse]
    total: int


class SavedIdItem(BaseModel):
    target_type: str
    target_id: str
